### PR TITLE
[ISSUE#5863] nacos server process batch register logic

### PR DESCRIPTION
**Pull request by chenhao26-nineteen**
Please do not create a Pull Request without creating an issue first.

## What is the purpose of the change

 For #5863 . nacos server process batch register logic 

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

* [ ] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
* [ ] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
* [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
* [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/alibaba/nacos/tree/master/test).
* [ ] Run `mvn -B clean package apache-rat:check findbugs:findbugs -Dmaven.test.skip=true` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.

### CODE REVIEW

File: naming/src/main/java/com/alibaba/nacos/naming/consistency/ephemeral/distro/v2/DistroClientDataProcessor.java [hunk 1]
```diff
@@ -166,16 +168,35 @@ private void upgradeClient(Client client, ClientSyncData clientSyncData) {
...
-                        new ClientOperationEvent.ClientRegisterServiceEvent(singleton, client.getClientId()));
+            InstancePublishInfo clientInstancePublishInfo = client.getInstancePublishInfo(singleton);
+            if (CollectionUtils.isNotEmpty(batchInstancePublishInfos)) {
+                BatchInstancePublishInfo batchInstancePublishInfo = batchInstancePublishInfos.get(i);
+                List<InstancePublishInfo> sourceInstanceInfo = batchInstancePublishInfo.getInstancePublishInfos();
+                BatchInstancePublishInfo targetInstanceInfo = (BatchInstancePublishInfo) clientInstancePublishInfo;
+                boolean result = false;
+                //If it is not empty, compare whether the instances are the same. If it is empty, notify directly
+                if (targetInstanceInfo != null) {
+                    result  = CollectionUtils.isEqualCollection(sourceInstanceInfo,
```
Comment (KomachiSion) [inline naming/src/main/java/com/alibaba/nacos/naming/consistency/ephemeral/distro/v2/DistroClientDataProcessor.java:185]: Why not override equals and hashcode?

File: naming/src/main/java/com/alibaba/nacos/naming/core/v2/index/ServiceStorage.java [hunk 2]
```diff
@@ -106,16 +108,43 @@ private List<Instance> getAllInstancesFromIndex(Service service) {
...
-                Instance instance = parseInstance(service, instancePublishInfo.get());
-                result.add(instance);
-                clusters.add(instance.getClusterName());
+                InstancePublishInfo publishInfo = instancePublishInfo.get();
+                //If it is a BatchInstancePublishInfo type, it will be processed manually and added to the instance list
+                if (publishInfo instanceof BatchInstancePublishInfo) {
+                    BatchInstancePublishInfo batchInstancePublishInfo = (BatchInstancePublishInfo) publishInfo;
+                    List<Instance> batchInstance = parseBatchInstance(service, batchInstancePublishInfo);
+                    result.addAll(batchInstance);
+                    for (Instance instance : batchInstance) {
```
Comment (KomachiSion) [inline naming/src/main/java/com/alibaba/nacos/naming/core/v2/index/ServiceStorage.java:117]: No need to do twice `for`

**Review by KomachiSion — CHANGES_REQUESTED** [review]
Please pay attention to the instance counts in `client` and `metricsMonitor` for batchRegister

File: naming/src/main/java/com/alibaba/nacos/naming/consistency/ephemeral/distro/v2/DistroClientDataProcessor.java (hunk 1 above)
Comment (chenhao26-nineteen) [reply naming/src/main/java/com/alibaba/nacos/naming/consistency/ephemeral/distro/v2/DistroClientDataProcessor.java:185]: This method can achieve the same purpose. It is also used to compare whether the fields of the internal elements of the set are equal. It has been tested

Comment (chenhao26-nineteen) [reply naming/src/main/java/com/alibaba/nacos/naming/consistency/ephemeral/distro/v2/DistroClientDataProcessor.java:185]: > Why not override equals and hashcode?

This method can achieve the same purpose. It is also used to compare whether the fields of the internal elements of the set are equal. It has been tested

File: naming/src/main/java/com/alibaba/nacos/naming/core/v2/index/ServiceStorage.java (hunk 2 above)
Comment (chenhao26-nineteen) [reply naming/src/main/java/com/alibaba/nacos/naming/core/v2/index/ServiceStorage.java:117]: > No need to do twice `for`

Why? ClusterName should be set here.

Comment (chenhao26-nineteen) [reply naming/src/main/java/com/alibaba/nacos/naming/core/v2/index/ServiceStorage.java]: > No need to do twice `for`

i fix it

File: naming/src/main/java/com/alibaba/nacos/naming/consistency/ephemeral/distro/v2/DistroClientDataProcessor.java (hunk 1 above)
Comment (chenhao26-nineteen) [reply naming/src/main/java/com/alibaba/nacos/naming/consistency/ephemeral/distro/v2/DistroClientDataProcessor.java]: > Why not override equals and hashcode?

i fix it

File: naming/src/main/java/com/alibaba/nacos/naming/consistency/ephemeral/distro/v2/DistroClientDataProcessor.java [hunk 3]
```diff
@@ -166,16 +168,33 @@ private void upgradeClient(Client client, ClientSyncData clientSyncData) {
...
             Service service = Service.newService(namespaces.get(i), groupNames.get(i), serviceNames.get(i));
             Service singleton = ServiceManager.getInstance().getSingleton(service);
             syncedService.add(singleton);
-            InstancePublishInfo instancePublishInfo = instances.get(i);
-            if (!instancePublishInfo.equals(client.getInstancePublishInfo(singleton))) {
-                client.addServiceInstance(singleton, instancePublishInfo);
-                NotifyCenter.publishEvent(
-                        new ClientOperationEvent.ClientRegisterServiceEvent(singleton, client.getClientId()));
+            InstancePublishInfo clientInstancePublishInfo = client.getInstancePublishInfo(singleton);
+            if (CollectionUtils.isNotEmpty(batchInstancePublishInfos)) {
```
Comment (KomachiSion) [inline naming/src/main/java/com/alibaba/nacos/naming/consistency/ephemeral/distro/v2/DistroClientDataProcessor.java:178]: What's about when synced batchInstancePublishInfos last time but no batchInstancePublishInfos next time?

**Review by KomachiSion — CHANGES_REQUESTED** [review]

File: naming/src/main/java/com/alibaba/nacos/naming/consistency/ephemeral/distro/v2/DistroClientDataProcessor.java (hunk 3 above)
Comment (chenhao26-nineteen) [reply naming/src/main/java/com/alibaba/nacos/naming/consistency/ephemeral/distro/v2/DistroClientDataProcessor.java:178]: > What's about when synced batchInstancePublishInfos last time but no batchInstancePublishInfos next time?

I fix it, the empty instance is judged in else, and the 400 status code is returned to the client in the handler

Comment (KomachiSion) [reply naming/src/main/java/com/alibaba/nacos/naming/consistency/ephemeral/distro/v2/DistroClientDataProcessor.java:178]: I think about it again, we not support the deregister batch, if we not allow users batchRegister empty list, how to remove these instance?

**Review by KomachiSion — CHANGES_REQUESTED** [review]

File: naming/src/main/java/com/alibaba/nacos/naming/core/v2/client/AbstractClient.java [hunk 4]
```diff
@@ -119,14 +121,41 @@ public ClientSyncData generateSyncData() {
...
+        return new ClientSyncData(getClientId(), namespaces, groupNames, serviceNames, instances, batchInstanceData);
+    }
+    
+    private static BatchInstanceData buildBatchInstanceData(BatchInstanceData  batchInstanceData, List<String> batchNamespaces,
+            List<String> batchGroupNames, List<String> batchServiceNames, Map.Entry<Service, InstancePublishInfo> entry) {
+        batchNamespaces.add(entry.getKey().getNamespace());
+        batchGroupNames.add(entry.getKey().getGroup());
+        batchServiceNames.add(entry.getKey().getName());
+        
+        batchInstanceData.setNamespaces(batchNamespaces);
```
Comment (KomachiSion) [inline naming/src/main/java/com/alibaba/nacos/naming/core/v2/client/AbstractClient.java:155]: duplicate operation

File: naming/src/main/java/com/alibaba/nacos/naming/core/v2/index/ServiceStorage.java [hunk 5]
```diff
@@ -107,16 +109,41 @@ private List<Instance> getAllInstancesFromIndex(Service service) {
...
             if (instancePublishInfo.isPresent()) {
-                Instance instance = parseInstance(service, instancePublishInfo.get());
-                result.add(instance);
-                clusters.add(instance.getClusterName());
+                InstancePublishInfo publishInfo = instancePublishInfo.get();
+                //If it is a BatchInstancePublishInfo type, it will be processed manually and added to the instance list
+                if (publishInfo instanceof BatchInstancePublishInfo) {
+                    BatchInstancePublishInfo batchInstancePublishInfo = (BatchInstancePublishInfo) publishInfo;
+                    List<Instance> batchInstance = parseBatchInstance(service, batchInstancePublishInfo, clusters);
+                    result.addAll(batchInstance);
```
Comment (KomachiSion) [inline naming/src/main/java/com/alibaba/nacos/naming/core/v2/index/ServiceStorage.java:117]: clusters？

File: naming/src/main/java/com/alibaba/nacos/naming/remote/rpc/handler/BatchInstanceRequestHandler.java [hunk 6]
```diff
@@ -26,6 +26,7 @@
 import com.alibaba.nacos.naming.core.v2.service.impl.EphemeralClientOperationServiceImpl;
 import org.springframework.stereotype.Component;
 
+
```
Comment (KomachiSion) [inline naming/src/main/java/com/alibaba/nacos/naming/remote/rpc/handler/BatchInstanceRequestHandler.java:29]: why add new line？ check style can pass？

**Review by KomachiSion — CHANGES_REQUESTED** [review]

File: naming/src/main/java/com/alibaba/nacos/naming/consistency/ephemeral/distro/v2/DistroClientDataProcessor.java [hunk 7]
```diff
@@ -187,6 +193,39 @@ private void upgradeClient(Client client, ClientSyncData clientSyncData) {
         }
     }
     
+    private static void processBatchInstanceDistroData(Set<Service> syncedService, Client client, ClientSyncData clientSyncData)  {
+        if (clientSyncData == null) {
```
[the remark this answers is not in the document]
Comment (chenhao26-nineteen) [reply naming/src/main/java/com/alibaba/nacos/naming/consistency/ephemeral/distro/v2/DistroClientDataProcessor.java:197]: Yes, the object cannot be null ，I will fix it

File: client/src/test/java/com/alibaba/nacos/client/naming/remote/gprc/NamingGrpcClientProxyTest.java [hunk 8]
```diff
@@ -331,7 +331,7 @@ public void close() {
         while (ORIGIN_SERVER.equals(rpc.getCurrentServer().getServerIp())) {
             TimeUnit.SECONDS.sleep(1);
             if (--retry < 0) {
-                Assert.fail("failed to auth switch server");
+                Assert.fail("failed to auth switch server！");
```
[the remark this answers is not in the document]
Comment (chenhao26-nineteen) [reply client/src/test/java/com/alibaba/nacos/client/naming/remote/gprc/NamingGrpcClientProxyTest.java:334]: Ci failed. It was modified to run Ci

File: config/src/test/java/com/alibaba/nacos/config/server/utils/GroupKey2Test.java [hunk 9]
```diff
@@ -38,23 +38,23 @@ public void testParseInvalidGroupKey2() {
             GroupKey2.parseKey(key);
             Assert.fail();
         } catch (IllegalArgumentException e) {
-            System.out.println(e.toString());
+            System.out.println(e);
```
[the remark this answers is not in the document]
Comment (chenhao26-nineteen) [reply config/src/test/java/com/alibaba/nacos/config/server/utils/GroupKey2Test.java:41]: > No this PR content.

Ci failed. It was modified to run Ci

File: console/src/main/resources/application.properties [hunk 10]
```diff
@@ -29,15 +29,15 @@ server.servlet.contextPath=/nacos
 
 #*************** Config Module Related Configurations ***************#
 ### If use MySQL as datasource:
-# spring.datasource.platform=mysql
-
+#spring.datasource.platform=mysql
+#
```
[the remark this answers is not in the document]
Comment (chenhao26-nineteen) [reply console/src/main/resources/application.properties:33]: > No this PR content.

Ci failed. It was modified to run Ci

File: naming/src/main/java/com/alibaba/nacos/naming/consistency/ephemeral/distro/v2/DistroClientDataProcessor.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (chenhao26-nineteen) [reply naming/src/main/java/com/alibaba/nacos/naming/consistency/ephemeral/distro/v2/DistroClientDataProcessor.java:197]: > Yes, the object cannot be null ，I will fix it

Yes, the object cannot be null ，I will fix it

**Review by KomachiSion — APPROVED** [review]

**Pull request merged**